### PR TITLE
Added recursion detection + unit tests

### PR DESCRIPTION
**Pull request by MacDada**

### CODE REVIEW

File: var_dump/_var_dump.py [hunk 1]
```diff
@@ -80,29 +80,35 @@ def dump(o, space, num, key, typ, proret):
 
     elif isinstance(o, object):
         r += display(o, space, num, key, typ, proret)
-        num = 0
```
Comment (sha256) [inline var_dump/_var_dump.py:83]: Thank you for the pull request. 

I think if you put
```
if recursion:
    r += ' …recursion…'
    return r
```
in this line, you could avoid nesting the whole code block.

Comment (sha256) [reply var_dump/_var_dump.py:83]: Also maybe a better wording for recursion is `circular reference` or `self reference`?

**Review by sha256 — CHANGES_REQUESTED** [review]
Thank you for the pull request. Please check my comments. And special thanks for the unit tests. This will speed up evaluating future pull requests from anyone.

File: var_dump/_var_dump.py (hunk 1 above)
Comment (MacDada) [reply var_dump/_var_dump.py:83]: > in this line, you could avoid nesting the whole code block.

Well, actually i made more refactorings later, but I didn't want to include them in the PR, as I wasn't sure if you're up to them:

* https://github.com/MacDada/python-var-dump/commits/hard_fork
* https://github.com/MacDada/python-var-dump/commit/b8c26e817f5e86f3d2b078a0bd7255030795c255
* https://github.com/MacDada/python-var-dump/commit/c23970ec05cb3b97ef72c5ae8146d970a9595730
* https://github.com/MacDada/python-var-dump/commit/7a9f48d664430220d2f7e56bb9f468af02403bda

Do you want me to include those?

Comment (MacDada) [reply var_dump/_var_dump.py:83]: > Also maybe a better wording for recursion is `circular reference` or `self reference`?

Yeah, I wasn't sure how to name it: for both the var name and [the output](https://github.com/sha256/python-var-dump/pull/22/files#diff-b6a56edd596ea041b5bcb04638c8fb85f039005cb38f07e0ff58f42400bfb688R131)

All 3 are good enough for me – name it ;)

Comment (sha256) [reply var_dump/_var_dump.py:83]: Awesome! This gives me an idea to go even further:
```
def dump(o, space, num, key, typ, proret, recursion=False):
    if type(o) in (str, int, float, long, bool, NoneType, unicode, Enum) or isinstance(o, Enum):
        return display(o, space, num, key, typ, proret)

    if not isinstance(o, object):
        return ''
    
    r = display(o, space, num, key, typ, proret)

    if recursion:
        return r + ' …recursion…'

    o_backup = o
    num = 0

    if type(o) in (tuple, list, dict):
        typ = type(o)  # type of the container of str, int, long, float etc
    elif isinstance(o, object):
        try:
            o = o.__dict__
        except AttributeError:
            return r
        typ = object
    for i in o:
        space += TAB_SIZE
        if type(o) is dict:
            r += dump(o[i], space, num, i, typ, proret, o[i] is o_backup)
        else:
            r += dump(i, space, num, '', typ, proret)
        num += 1
        space -= TAB_SIZE
    return r
```

Comment (MacDada) [reply var_dump/_var_dump.py:83]: > This gives me an idea to go even further:

https://github.com/sha256/python-var-dump/pull/22/commits/8cfe28b4d0b6a4cdd9792803c6569fd5e48b0360


> Please go with ` …circular reference…`

https://github.com/sha256/python-var-dump/pull/22/commits/61e3b8f32456a588b737f899c63c62948fcab49e

**Review by sha256 — APPROVED** [review]

**Comment by MacDada** [comment]
Thanks! <3

**Pull request merged**